app/services/metrics/review_turnaround_per_project_processor.rb

```diff
@@ -0,0 +1,120 @@
+  class ReviewTurnaroundPerProjectProcessor < BaseMetricProcessor
+    private
+
+    ##
+    # review_turnaround_per_project: { project_id: review_turnaround_value }
+    #   Keeps the values of the review_turnaround for each project.
+    #   Allows to calculate the metrics for all the projects in a single pass.
+    #
+    # pull_request_reviewed: { pull_request_id: true|nil }
+    #   A PullRequest can have many Review events. The
+    #   review_turnaround only considers the first Review. This flag is set
+    #   to true if a previous Review event was the first Review to ignore all
+    #   the Review events after that one.
+    def create_accumulators
+      {
+        review_turnaround_per_project: Hash.new { |hash, key| hash[key] = [] },
+        pull_request_reviewed: {}
+      }
+    end
+
+    def process_event(event:, accumulators:)
+      pull_request_id = event.data['pull_request']['id']
+
+      review_turnaround_value = review_turnaround_as_seconds(event: event)
+      accumulators[:review_turnaround_per_project][event.project.name] << review_turnaround_value
+      accumulators[:pull_request_reviewed][pull_request_id] = :reviewed
+    end
+
+    ##
+    # Updates the metrics for all the project in the given review_turnaround_per_project
+    def update_metrics(time_interval:, accumulators:)
+      accumulators[:review_turnaround_per_project].each_pair do |project_id, values|
+        average_value = values.sum / values.size
+
+        update_metric(
+          entity_key: project_id,
+          metric_key: 'review_turnaround',
+          value: average_value,
+          value_timestamp: time_interval.starting_at
+        )
+      end
+    end
+
+    ##
+    # Returns the review turnaround value for the given event.
+    def review_turnaround_as_seconds(event:)
+      payload = event.data
+
+      # RFC should we move these times to the Event model?
+      reviewed_at = parse_time(payload['review']['submitted_at'])
+      review_requested_at = parse_time(payload['pull_request']['created_at'])
+
+      (reviewed_at - review_requested_at).seconds
+    end
+
+    def parse_time(time_string)
+      Time.zone.parse(time_string)
+    end
+
+    ##
+    # The review turnaround metric is only interested in the first review of a PR.
+    # This method returns true if the PR of the given review_event was already reviewed.
+    #
+    # A review_event is the first review of a PR if:
+    #     - its action = 'submitted'
+    #       &&
+    #     - its PR has exactly one Review (the one created by this very same review_event)
+    #
+    # In all other cases the pull_request linked to this review_event was already
+    # reviewed and the review_event is ignored for the review_turnaround metric
+    def skip_event?(event:, time_interval:, accumulators:)
+      payload = event.data
+      event.name != 'review' ||
+        payload['action'] != 'submitted' ||
+        accumulators[:pull_request_reviewed].key?(payload['pull_request']['id']) ||
```

Comment:
so you are skipping the event if the `accumulators[:pull_request_reviewed]` already has the project key. But how did you know you needed to process the previous event and not this one? What guaranteed the order?